### ISSUE TITLE
DTO / ObjectWrapper Declaration Syntax

### ISSUE THREAD

**Issue by lganzzzo** [oatpp/oatpp#247]
# DTO / ObjectWrapper Declaration Syntax

**Everyone is welcome to comment on this one.**  
*(You may just leave comments in form - "v1" or "v2". Additional comments are optional.)*

---

## Variant 1

```cpp
class MyDto : public oatpp::Object {

  DTO_INIT(MyDto, Object)

  DTO_FIELD(MyDto::Wrapper, myDto);
  DTO_FIELD(List<MyDto::Wrapper>, listOfDto);

}

...

MyDto::Wrapper myDto;
oatpp::List<MyDto::Wrapper> listOfDto;
```

---

## Variant 2

```cpp
class MyDto : public oatpp::DTO {

  DTO_INIT(MyDto, DTO)

  DTO_FIELD(Object<MyDto>, myDto);
  DTO_FIELD(List<Object<MyDto>>, listOfDto);

}

...

oatpp::Object<MyDto> myDto;
oatpp::List<oatpp::Object<MyDto>> listOfDto;
```

**Comment by nicraMarcin** [comment]
v1

**Comment by bhorn** [comment]
Definitely voting for V2. It's much easier to understand what is going on: `oatpp::Object<MyDto>`.
It's an oatpp-object of type MyDto. 
`oatpp::List<oatpp::Object<MyDto>>` is a list of objects of type `MyDto`.

While in V1, It's a bit confusing. One is `oatpp::XXX` the other is `XXX::Wrapper`. And what exactly is `::Wrapper`? It's not clear without further studies to new developers.

V2 is much clearer, it directly shows the intention **this** is an Object of **that**. Even when not knowing the codebase, you can easily understand the purpose or intention.

**Comment by quant86** [comment]
v2

**Comment by lganzzzo** [comment]
Thanks, everyone for voting, and for constructive comments!
The v3 was implemented - (the v2 + optional `::Wrapper` thing):

```cpp
class MyDto : public oatpp::DTO {

  DTO_INIT(MyDto, DTO)

  DTO_FIELD(Object<MyDto>, myDto);            // <--- Recommended syntax
  DTO_FIELD(List<Object<MyDto>>, listOfDto);  // <--- Recommended syntax

  DTO_FIELD(MyDto::Wrapper, myDto);           // <--- Will also work. NON-Recommended syntax
  DTO_FIELD(List<MyDto::Wrapper>, listOfDto); // <--- Will also work. NON-Recommended syntax

}

...

oatpp::Object<MyDto> myDto;                   // <--- Recommended syntax
oatpp::List<oatpp::Object<MyDto>> listOfDto;  // <--- Recommended syntax

MyDto::Wrapper myDto;                         // <--- Will also work. NON-Recommended syntax
oatpp::List<MyDto::Wrapper> listOfDto;        // <--- Will also work. NON-Recommended syntax
```

**Comment by lganzzzo** [comment]
This ticket won't be closed up until the next (after `1.1.0`) release.
Feel free to add comments and propositions.

**Comment by lganzzzo** [comment]
I'm closing this issue.
The new API is considered to be stable.
Thanks to everyone!
